locator/token_metadata.cc

```diff
@@ -769,122 +719,113 @@ token_metadata_impl::has_pending_ranges(sstring keyspace_name, inet_address endp
-        auto current_endpoints = strategy.calculate_natural_endpoints(t, metadata).get0();
-        auto new_endpoints = strategy.calculate_natural_endpoints(t, *all_left_metadata).get0();
-        for (auto ep : new_endpoints) {
-          if (!current_endpoints.contains(ep)) {
-            new_pending_ranges.emplace(r, ep);
-          }
-        }
-        seastar::thread::maybe_yield();
-    }
-    metadata.clear_gently().get();
-    tlogger.debug("In calculate_pending_ranges: affected_ranges.size={} ends", affected_ranges_size);
-}
-
-// Called from a seastar thread
-void token_metadata_impl::calculate_pending_ranges_for_replacing(
-        const token_metadata& unpimplified_this,
-        const abstract_replication_strategy& strategy,
-        std::unordered_multimap<range<token>, inet_address>& new_pending_ranges) const {
-    if (_replacing_endpoints.empty()) {
-        return;
-    }
-    for (const auto& node : _replacing_endpoints) {
-        auto existing_node = node.first;
-        auto replacing_node = node.second;
-        auto address_ranges = strategy.get_ranges(existing_node, unpimplified_this).get0();
-        for (const dht::token_range& x : address_ranges) {
-            seastar::thread::maybe_yield();
-            tlogger.debug("Node {} replaces {} for range {}", replacing_node, existing_node, x);
-            new_pending_ranges.emplace(x, replacing_node);
-        }
-    }
-}
-
-// Called from a seastar thread
-void token_metadata_impl::calculate_pending_ranges_for_bootstrap(
-        const abstract_replication_strategy& strategy,
-        std::unordered_multimap<range<token>, inet_address>& new_pending_ranges,
-        mutable_token_metadata_ptr all_left_metadata, dc_rack_fn& get_dc_rack) const {
-    // For each of the bootstrapping nodes, simply add and remove them one by one to
-    // allLeftMetadata and check in between what their ranges would be.
-    std::unordered_multimap<inet_address, token> bootstrap_addresses;
-    for (auto& x : _bootstrap_tokens) {
-        bootstrap_addresses.emplace(x.second, x.first);
-    }
-
-    // TODO: share code with unordered_multimap_to_unordered_map
-    std::unordered_map<inet_address, std::unordered_set<token>> tmp;
-    for (auto& x : bootstrap_addresses) {
-        auto& addr = x.first;
-        auto& t = x.second;
-        tmp[addr].insert(t);
-    }
-    for (auto& x : tmp) {
-        auto& endpoint = x.first;
-        auto& tokens = x.second;
-        all_left_metadata->update_topology(endpoint, get_dc_rack(endpoint));
-        all_left_metadata->update_normal_tokens(tokens, endpoint).get();
-        auto address_ranges = strategy.get_ranges(endpoint, *all_left_metadata).get0();
-        for (const dht::token_range& x : address_ranges) {
-            new_pending_ranges.emplace(x, endpoint);
-        }
-        all_left_metadata->_impl->remove_endpoint(endpoint);
-    }
-    all_left_metadata->_impl->sort_tokens();
-}
-
 future<> token_metadata_impl::update_pending_ranges(
```

Comment:
> total: O((n + m) * (replication_factor + log(n + m))).

The `n*m` from first bullet is missing.

Although I don't understand where the `n*m` comes from here:
> new_token_metadata: O(n * m + (n + m)log(n+m)), multiplication is from inefficient linear scans in update_normal_tokens, log is from sort_tokens

IIUC `update_normal_tokens` itself is `(n+m)log(n+m)`. Are you multiplying by `n` because of this loop?
```
            // apply new_normal_tokens
            for (auto& [endpoint, tokens]: new_normal_tokens) {
                result->update_topology(endpoint,
                    get_dc_rack(endpoint),
                    _normal_token_owners.contains(endpoint) ? node::state::normal : node::state::joining);
                result->update_normal_tokens(std::move(tokens), endpoint).get();
            }
```
but (at least today) there should be at most one pair in `new_normal_tokens`, so there should be a single call to `update_normal_tokens` with `tokens.size() == n`. So it should be just `(n+m)log(n+m)` from the single call to `update_normal_tokens`.

If the new tokens were spread among multiple endpoints the complexity analysis seems more complex. If we naively estimated `new_normal_tokens.size() = O(n)` then we would get `O(n*(n+m)log(n+m))`.

Reply:
I don't feel much value in precise O-estimation of the two approaches.

The important thing (mentioned in the refactoring commit message) is that old approach called `get_ranges` for each affected node. `get_ranges` is basically `calculate_natural_endpoints` for each token. So, old approach did that at least once, new approach - exactly only -> it's not worse than the old.